src/nix-store/nix-store.cc

```diff
@@ -950,8 +950,12 @@ static void opServe(Strings opFlags, Strings opArgs)
                 info.sigs = readStrings<StringSet>(in);
                 in >> info.ca;
 
-                // FIXME: race if addToStore doesn't read source?
-                store->addToStore(info, in, NoRepair, NoCheckSigs);
+                SizedSource sizedSource(in, info.narSize);
+
+                store->addToStore(info, sizedSource, NoRepair, NoCheckSigs);
+
+                // consume all the data that has been sent before continuing.
+                sizedSource.drainAll();
```

Comment:
Won't this block indefinitely waiting for EOF?

Reply:
In any case I'd except this to be something like `sizedSource.discard(sizedSource.remain)`.

Reply:
`drainAll()` discards all the data until it reaches EOF. SizedSource emits EOF when `remain` is zero. So it does what you expects but uses an exception as part of the control flow.

What I could do is attack drainAll() to sizedSource instead and be a bit smarter about this.

Reply:
in f21662b I now implement `drainAll()` on the SizedSource only (since it's the only user right now) and which allows me to `read()` until `remain` is zero. And avoid using the Except for control flow.